packages/eslint-config-compass/index.js

```diff
@@ -5,6 +5,12 @@ const tsConfigurations = [
   "plugin:@typescript-eslint/recommended",
   "plugin:@typescript-eslint/recommended-requiring-type-checking",
 ];
+const tsRules = {
+  "@typescript-eslint/no-unsafe-assignment": "off",
+  "@typescript-eslint/no-unsafe-call": "off",
+  "@typescript-eslint/no-unsafe-member-access": "off",
+  "@typescript-eslint/no-unsafe-return": "off",
+};
```

Comment:
I would prefer that we don't do this for the all repositories by default, those are good rules I think 😄 If we temporarily need to have them disabled in data-service, maybe let's just do an override there with a comment explaining why and a plan/todo ticket reference for how we can turn them back on, what do you think?

Reply:
The problem for those is basically that you interact with the driver in quite a couple of places. As the result of those is always a `Document` (which gets inferred to as `any` in the end), you have those issues all over the place. Also, as long as we are migrating towards TypeScript, we are forced to use `any` in a lot of places unfortunately. As such, all these locations would require us to manually turn those off. I'd rather suggest to keep them of until the biggest part of the code base is migrated so we can have proper types at all locations.

Reply:
I see your point about the driver returning very generic types, but I really don't think that falling back to `any` is something that we should just accept and proceed with it if we want to get most of the benefits of using typescript in our code. _I really don't want to block this particular migration work_ also, so this is why I'm suggesting to disable those only on the workspace/package level with a ticket in the backlog.

As for the way forward that I would suggest I'll give an example of how I see it and you tell me what you think and if it makes sense (if it doesn't then yes, let's keep everything as is). So to make the example more realistic I pulled the branch and turned on the rules back just to get a sense of how bad it is, and, yes, I can see that especially `instance-detail-helper` lit up with red as a Christmas tree, let's go through the files now.

- `data-service.ts` is all good, skipping
- `native-client.ts` has some errors because of those rules, let's dig in deeper

So some of those any warnings are due to tunnel and connection model not having any types, I think we should provide the types there for the stuff we expect those to have as suggested in https://github.com/mongodb-js/compass/pull/2406#discussion_r691954683 and https://github.com/mongodb-js/compass/pull/2406#discussion_r691954683 (even better I think if we provide d.ts with light connection model interface typed for the time being while the `connection-model` package is not in typescript, this is also relevant for some other packages that we import in the code there)

Let's look at the `Document` usage that causes issues now. The first one by order in the code was this method:

```ts
  _buildCollectionDetail(ns: string, data: Document): Document {
    return assignIn<Document>(data.stats, {
      _id: ns,
      name: this._collectionName(ns),
      database: this._databaseName(ns),
      indexes: data.indexes,
      //           ^ typescript is not happy with this property access
    });
```

So in this case `_buildCollectionDetail` is an internal method, so why can't we just type it without a Document type? It's not even really just receiving those options straight from the driver, this is what we pass to it, so let's fix it:

```ts
  _buildCollectionDetail(ns: string, data: { indexes?: unknown, stats?: Document }): Document {
  //                                                   ^ we can do better here, but let's say we don't care for now (or at all if it's actually unknown)
```

Now let's take care of the return type, this method merges some additional props to collstats and returns it. Driver has collstats types provided, so we can just use them there, right? This is how it all can look all together:

```ts
  _buildCollectionDetail(
    ns: string,
    data: { indexes?: unknown; stats?: CollStats }
  ): CollStats & {
    _id: string;
    name: string;
    database: string;
    indexes: unknown;
  } {
    return assignIn(data.stats, {
      _id: ns,
      name: this._collectionName(ns),
      database: this._databaseName(ns),
      indexes: data.indexes,
    });
  }
```

There is now no eslint errors and we have some type contracts in code for what exactly we are expecting to be provided in the method and what it returns.

Taking into account that data-service is our layer between driver and compass (at least for now) it think it would be really-really helpful if we could avoid leaking those `Document` types outside where possible. To give another example, we have a `collectionStats` method that currently returns `Document`, but we know exactly what it returns (a result of `_buildCollectionStats` call that is in our code, not in driver), so I think we probably should type it.

I'll stop with the examples here because I don't want to blow up the size of this comment even more, but if you have some good ones that you would want to take a look at from my point of view, please tell me!

Also just want to note that yes, there will be cases where we would still hit unsolvable issues with `any` type being involved, but there should not be so many that we have to disable those rules for the whole codebase, those should be exceptions skipped with `eslint-ignore <rulename>` or `ts-expect-error`.

Is this a lot of work to type all those pieces where we interact with the driver? Well, in my opinion it's not impossible, but yes, it's not a small amount of work also. But I think it's an important one that we should not skip on! And this is why I'm suggesting to scope the rule overrides to this package and resolve this when we can.

I know that you already put a lot of thought into it and again, all this work you did looks amazing! So please, if I'm missing something big here or just saying some nonsense just say it right away

Reply:
Okay I've tried doing it but it is a huge pain :D And I found out that a lot of the stuff... honestly is so outdated that it just doesn't exist anymore in the documents. E.g. there just is no `shards` field returned when running `collStats` 🤷 So I'd like to encapsulate that as cleanup in a separate PR and I'll follow up with that immediately afterwards. But unfortunately I can't type that stuff since it basically doesn't exist and isn't used either :D In order to limit those changes and make those more clear - I'd suggest doing that separately 👍

Reply:
Totally makes sense, yeah! I appreciate all the additional work you already put into this, this patch now looks even better than before if that's even possible 🤩